### DISCUSSION TITLE
Start from a block selection, then expand by word?

### DISCUSSION THREAD

**Discussion by ktfleming** [meow-edit/meow#118] [Q&A] [answered]
Great work on Meow! I've been really impressed with trying it out so far.

There's one thing I haven't been able to do, and I'm not sure if I'm just missing how to do it, or if it's not possible. Suppose I have the following JS code:

```js
const foo = bar({
  one,
  two,
  three
});
```

My cursor is on `two`, and I hit the `o` key to select the block (I'm using the recommended QWERTY mappings, so `meow-block`.). Now I've selected around the curly braces. I hit `o` again to expand, and now I've selected around the parentheses. Everything is good so far.

Now, I'd like to expand the selection backward by one word, so I want to add `bar` to my selection. I was hoping that pressing `b` (`meow-back-word`) would do so, but instead it seems like it cancels the selection and starts a new one.

Is there a way to do this in Meow? I've used Vim/Evil for a while so I'm used to that style of modal editing, but I'm new to the Meow style, so it's possible I'm missing something, or thinking about it in the wrong way.

**Comment by Linerre** [comment, marked as answer]
Hello, @ktfleming

Here is my suggestion:
After you've selected all the parentheses and its included content:
- press `;` (`meow-reverse`) to move your cursor to the beginning of the selection
- press `Shift` + `h` to activate `meow-left-expand`, then just press `h` to select a single char each time.

There is another way to achieve the same effect:
- move your cursor to letter "*b*" of "bar"
- press `Shift` + `o`
Done! 
The uppercase `O` relates to `meow-to-block` command. It will select from your current position all the way to the first block (inclusive) it hits.

The developer may have better solutions, so stay tuned.

**Reply by DogLooksGood** [reply]
In this case, assuming your cursor is at the `w` in word `two`, I prefer `b(back-word) 2(expand-2) O(to-block)`.

**Reply by ktfleming** [reply]
I see! That works -- I'll experiment more with`meow-to-block`. Thanks!

**Reply by DogLooksGood** [reply]
Also thing-based command is extensible, and I just finished the refactor for that part. 
Now you can register something like
```
(meow-thing-register 'extend '(syntax . "^-><().") '(syntax . "^-><()"))
```
And add it to the `meow-char-thing-table`
```
(add-to-list 'meow-char-thing-table '(?x . extend))
```

Then you can use `o o , x`.

**Reply by fbergroth** [reply]
Coming from many years of evil-mode, I also got a bit confused by a similar situation.

The answers above are good, but I see two problems.

> * press `;` (`meow-reverse`) to move your cursor to the beginning of the selection
> * press `Shift` + `h` to activate `meow-left-expand`, then just press `h` to select a single char each time.

I think this is a good solution, but I would really like to use something like `meow-back-word` that extends backwards instead of character by character. 

> In this case, assuming your cursor is at the `w` in word `two`, I prefer `b(back-word) 2(expand-2) O(to-block)`.

This is very elegant, but I think my brain works the other way around. I would like to first select the block, and then expand the region to include anything that was missed. Maybe it's something you'll get used to with meow.

I think I would prefer having `e` was `meow-next-word`, and `E` `meow-next-word-expand` (and same for `b / B`,  `meow-back-word/-expand` ), similarly to hjkl / HJKL, but maybe I just haven't grokked meow yet :-)

**Reply by DogLooksGood** [reply]
The idea is simple, when you want to go(select) to place X, the only thing you should care is where is X. Is X the boundary of block? Is X a end of a word? Or is X just after a character `a`? You care about the target place, not the path you take to get there.

If you want to select from A to B. 
- You go to A
- Move towards to B by the thing you describe what B is
- Expanding
